File: be/test/runtime/s3_plugin_downloader_test.cpp
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

#include "runtime/s3_plugin_downloader.h"

#include <gtest/gtest.h>

#include <filesystem>
#include <string>

namespace doris {

class S3PluginDownloaderTest : public ::testing::Test {
public:
    S3PluginDownloaderTest()
            : s3_config_("s3.test-endpoint.com", "us-west-2", "test-bucket", "ak", "sk") {}

    void SetUp() override {
        test_dir_ = std::filesystem::temp_directory_path() / "s3_plugin_test";
        std::filesystem::create_directories(test_dir_);
        local_path_ = (test_dir_ / "test-plugin.jar").string();
    }

    void TearDown() override {
        if (std::filesystem::exists(test_dir_)) {
            std::filesystem::remove_all(test_dir_);
        }
    }

protected:
    std::filesystem::path test_dir_;
    S3PluginDownloader::S3Config s3_config_;
    std::string local_path_;
};

TEST_F(S3PluginDownloaderTest, TestS3ConfigCreation) {
    EXPECT_EQ(s3_config_.endpoint, "s3.test-endpoint.com");
    EXPECT_EQ(s3_config_.region, "us-west-2");
    EXPECT_EQ(s3_config_.bucket, "test-bucket");
    EXPECT_EQ(s3_config_.access_key, "ak");
    EXPECT_EQ(s3_config_.secret_key, "sk");
}

TEST_F(S3PluginDownloaderTest, TestS3ConfigToString) {
    std::string expected_str =
            "S3Config{endpoint='s3.test-endpoint.com', region='us-west-2', "
            "bucket='test-bucket', access_key='***'}";
    EXPECT_EQ(s3_config_.to_string(), expected_str);

    S3PluginDownloader::S3Config config_no_ak("s3.test-endpoint.com", "us-west-2", "test-bucket",
                                              "", "sk");
    std::string expected_str_no_ak =
            "S3Config{endpoint='s3.test-endpoint.com', region='us-west-2', "
            "bucket='test-bucket', access_key='null'}";
    EXPECT_EQ(config_no_ak.to_string(), expected_str_no_ak);
}

} // namespace doris

File: fe/fe-core/src/test/java/org/apache/doris/common/plugin/S3PluginDownloaderTest.java
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.apache.doris.common.plugin;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Path;

public class S3PluginDownloaderTest {

    @Test
    public void testS3ConfigToString() {
        S3PluginDownloader.S3Config config = new S3PluginDownloader.S3Config(
                "", "us-west-2", "bucket", "prefix", "access-key", "secret");
        String configStr = config.toString();
        Assertions.assertTrue(configStr.contains("***"));
        Assertions.assertTrue(configStr.contains("prefix='prefix'"));

        S3PluginDownloader.S3Config emptyKeyConfig = new S3PluginDownloader.S3Config(
                "", "us-west-2", "bucket", "prefix", null, "secret");
        String emptyStr = emptyKeyConfig.toString();
        Assertions.assertTrue(emptyStr.contains("null"));
    }

    @Test
    public void testS3ConfigConstructor() {
        S3PluginDownloader.S3Config config = new S3PluginDownloader.S3Config(
                "", "us-west-2", "bucket", "prefix", "key", "secret");
        Assertions.assertEquals("", config.endpoint);
        Assertions.assertEquals("us-west-2", config.region);
        Assertions.assertEquals("bucket", config.bucket);
        Assertions.assertEquals("prefix", config.prefix);
        Assertions.assertEquals("key", config.accessKey);
        Assertions.assertEquals("secret", config.secretKey);

        // Test constructor doesn't throw
        Assertions.assertDoesNotThrow(() -> {
            new S3PluginDownloader(config);
        });
    }

    @Test
    public void testDownloadFileFailure() {
        S3PluginDownloader.S3Config config = new S3PluginDownloader.S3Config(
                "", "", "", "", "", "");
        S3PluginDownloader downloader = new S3PluginDownloader(config);

        RuntimeException exception = Assertions.assertThrows(RuntimeException.class, () -> {
            downloader.downloadFile("s3://bucket/file.jar", "/tmp/test.jar");
        });
        Assertions.assertTrue(exception.getMessage().contains("Download failed"));
    }

    @Test
    public void testAutoCloseable() throws Exception {
        S3PluginDownloader.S3Config config = new S3PluginDownloader.S3Config(
                "", "", "", "", "", ""); // Use empty config to avoid endpoint validation

        Assertions.assertDoesNotThrow(() -> {
            try (S3PluginDownloader downloader = new S3PluginDownloader(config)) {
                // Should not throw during close
            }
        });
    }


    @Test
    public void testDirectoryCreation(@TempDir Path tempDir) {
        S3PluginDownloader.S3Config config = new S3PluginDownloader.S3Config(
                "", "", "", "", "", ""); // Use empty config to avoid endpoint validation
        S3PluginDownloader downloader = new S3PluginDownloader(config);

        // Test directory creation by using nested path
        String nestedPath = tempDir.resolve("nested/dir/test.jar").toString();

        RuntimeException exception = Assertions.assertThrows(RuntimeException.class, () -> {
            downloader.downloadFile("s3://bucket/file.jar", nestedPath);
        });

        // Should fail at S3 download but directory creation should work
        Assertions.assertTrue(exception.getMessage().contains("Download failed"));

        // Verify directory was created
        Assertions.assertTrue(tempDir.resolve("nested/dir").toFile().exists());
    }

    @Test
    public void testExistingFileHandling(@TempDir Path tempDir) throws IOException {
        S3PluginDownloader.S3Config config = new S3PluginDownloader.S3Config(
                "", "", "", "", "", ""); // Use empty config to avoid endpoint validation
        S3PluginDownloader downloader = new S3PluginDownloader(config);

        // Create a test file
        File testFile = tempDir.resolve("existing-file.jar").toFile();
        try (FileWriter writer = new FileWriter(testFile)) {
            writer.write("test content");
        }

        Assertions.assertTrue(testFile.exists());

        // Attempt download - should delete existing file first
        RuntimeException exception = Assertions.assertThrows(RuntimeException.class, () -> {
            downloader.downloadFile("s3://bucket/file.jar", testFile.getAbsolutePath());
        });

        // Should fail at S3 download, but file deletion logic should be exercised
        Assertions.assertTrue(exception.getMessage().contains("Download failed"));
    }

}
